Source/DafnyCore/Resolver/ExpectContracts.cs

```diff
@@ -0,0 +1,291 @@
+  }
+
+  private static Expression MakeApplySuffix(IToken tok, string name, List<Expression> args) {
+    var nameExpr = new NameSegment(tok, name, null);
+    var argBindings = args.ConvertAll(arg => new ActualBinding(null, arg));
+    return new ApplySuffix(tok, null, nameExpr, argBindings, tok);
+  }
+
+  private bool ShouldGenerateWrapper(MemberDecl decl) {
+    return !decl.IsGhost &&
+           decl is not Constructor &&
+           CallRedirector.HasExternAttribute(decl);
+  }
+
+  /// <summary>
+  /// Create a wrapper for the given function or method declaration that
+  /// dynamically checks all of its preconditions, calls it, and checks
+  /// all of its postconditions before returning. Then add the new wrapper
+  /// as a sibling of the original declaration.
+  /// </summary>
+  /// <param name="parent">The declaration containing the on to be wrapped.</param>
+  /// <param name="decl">The declaration to be wrapped.</param>
+  private void GenerateWrapper(TopLevelDeclWithMembers parent, MemberDecl decl) {
+    var tok = decl.tok;
+
+    var newName = decl.Name + "__dafny_checked";
+    MemberDecl newDecl = null;
+
+    if (decl is Method origMethod) {
+      var newMethod = cloner.CloneMethod(origMethod);
+      newMethod.Name = newName;
+
+      var args = newMethod.Ins.Select(Expression.CreateIdentExpr).ToList();
+      var outs = newMethod.Outs.Select(Expression.CreateIdentExpr).ToList();
+      var applyExpr = MakeApplySuffix(tok, origMethod.Name, args);
+      var applyRhs = new ExprRhs(applyExpr);
+      var callStmt = new UpdateStmt(tok, tok, outs, new List<AssignmentRhs>() { applyRhs });
+
+      var body = MakeContractCheckingBody(origMethod.Req, origMethod.Ens, callStmt);
+      newMethod.Body = body;
+      newDecl = newMethod;
+    } else if (decl is Function origFunc) {
+      var newFunc = cloner.CloneFunction(origFunc);
+      newFunc.Name = newName;
+
+      var args = origFunc.Formals.Select(Expression.CreateIdentExpr).ToList();
+      var callExpr = MakeApplySuffix(tok, origFunc.Name, args);
+      newFunc.Body = callExpr;
+
+      var localName = origFunc.Result?.Name ?? "__result";
+      var local = new LocalVariable(tok, tok, localName, origFunc.ResultType, false);
+      var localExpr = new IdentifierExpr(tok, localName);
+      var callRhs = new ExprRhs(callExpr);
+
+      var lhss = new List<Expression> { localExpr };
+      var locs = new List<LocalVariable> { local };
+      var rhss = new List<AssignmentRhs> { callRhs };
+
+      var callStmt = origFunc.Result?.Name is null
+        ? (Statement)new VarDeclStmt(tok, tok, locs, new UpdateStmt(tok, tok, lhss, rhss))
+        : (Statement)new UpdateStmt(tok, tok, lhss, rhss);
+
+      var body = MakeContractCheckingBody(origFunc.Req, origFunc.Ens, callStmt);
+
+      if (origFunc.Result?.Name is null) {
+        body.AppendStmt(new ReturnStmt(tok, tok, new List<AssignmentRhs> { new ExprRhs(localExpr) }));
+      }
+      newFunc.ByMethodBody = body;
```

Comment:
Clever ! You transform the ByMethod body of an extern function with these expect statements. Love it.